File: src/clang/magma/compile/rule/text/InfixRule.c
#include "InfixRule.h"
int __lambda0__(){return right;
}
int __lambda1__(){return tuple;
}
public InfixRule(Rule left, String infix, Rule right, Locator locator){this.left = left;
        this.infix = infix;
        this.right = right;
        this.locator = locator;
}
Result<Node, CompileError> parse(String input){return locator.locate(input, infix).map(index -> {
            String left = input.substring(0, index);
            String right = input.substring(index + infix.length());
            return this.left.parse(left).and(() -> this.right.parse(right)).mapValue(tuple -> tuple.left().merge(tuple.right()));
        }).orElseGet(() -> {
            return new Err<>(new CompileError("Infix '" + infix + "' not present", new StringContext(input)));
        });
}
Result<String, CompileError> generate(Node node){return left.generate(node).and(__lambda0__.generate(node)).mapValue(__lambda1__.left()+infix+tuple.right());
}

File: src/java/magmac/compile/InfixRule.java
package magmac.compile;

import java.util.Optional;

public record InfixRule(
        Rule leftRule,
        String infix,
        Rule rightRule
) implements Rule {
    @Override
    public Optional<MapNode> parse(String input) {
        var index = input.indexOf(this.infix);
        if (0 > index) {
            return Optional.empty();
        }

        var leftString = input.substring(0, index);
        var rightString = input.substring(index + this.infix.length());
        return this.leftRule.parse(leftString).flatMap(leftValue -> {
            return this.rightRule.parse(rightString).map(rightValue -> {
                return leftValue.merge(rightValue);
            });
        });
    }

    @Override
    public Optional<String> generate(MapNode node) {
        return this.leftRule.generate(node).flatMap(leftString -> {
            return this.rightRule.generate(node).map(rightString -> {
                return leftString + this.infix + rightString;
            });
        });
    }
}